File: scripts/state-parks-to-sql.py
import csv
import re

input_file = "./scripts/active_state_parks.csv"
output_file = "./scripts/insert_state_park_badges.sql"

def escape_sql(text):
    """Escape single quotes for SQL."""
    return text.replace("'", "''")

def add_state_park_suffix(name):
    """Append 'State Park' if not already present at the end."""
    return name if name.strip().lower().endswith("state park") else f"{name} State Park"

def clean_id_for_sql(name, state_code):
    """Generate a safe, unique ID for SQL."""
    # Remove 'State Park' if it exists at the end
    base_name = re.sub(r'\s*state park$', '', name.strip(), flags=re.IGNORECASE)
    # Lowercase and replace non-alphanumeric characters with underscores
    safe_name = re.sub(r'\W+', '_', base_name.lower())
    # Append state code and "_state_park" suffix
    return f"{safe_name}_{state_code}_state_park"

with open(input_file, newline='', encoding='utf-8') as csv_in, \
     open(output_file, 'w', encoding='utf-8') as sql_out:

    reader = csv.DictReader(csv_in)
    count = 0
    for row in reader:
        name = row['name']
        # Extract state code from locationDesc (US-PA → PA)
        state_code = row['locationDesc'].split('-')[-1].lower()

        # Generate unique ID
        park_id = clean_id_for_sql(name, state_code)

        # Generate title and description
        title = escape_sql(add_state_park_suffix(name))
        description = escape_sql(f"completed 2 trails and visited {add_state_park_suffix(name)}")

        # Static icon and points
        icon = "state_park.svg"
        points = 0

        # Write INSERT statement
        sql_line = f"INSERT INTO badges (id, title, description, icon_svg, points) VALUES ('{park_id}', '{title}', '{description}', '{icon}', {points});\n"
        sql_out.write(sql_line)
        count += 1

print(f"SQL script generated safely with {count} active state parks: {output_file}")

File: scripts/state-parks-to-sql.js
/**
 * generateStateParkInserts_fixed.js
 * Node.js script to scrape Wikipedia lists of state parks
 * and generate SQL INSERT statements for your badges table.
 */

import fetch from "node-fetch";
import * as cheerio from "cheerio";
import fs from "fs";

// 50 US states and their abbreviations
const states = [
  { name: "Alabama", abbr: "AL" },
  { name: "Alaska", abbr: "AK" },
  { name: "Arizona", abbr: "AZ" },
  { name: "Arkansas", abbr: "AR" },
  { name: "California", abbr: "CA" },
  { name: "Colorado", abbr: "CO" },
  { name: "Connecticut", abbr: "CT" },
  { name: "Delaware", abbr: "DE" },
  { name: "Florida", abbr: "FL" },
  { name: "Georgia", abbr: "GA" },
  { name: "Hawaii", abbr: "HI" },
  { name: "Idaho", abbr: "ID" },
  { name: "Illinois", abbr: "IL" },
  { name: "Indiana", abbr: "IN" },
  { name: "Iowa", abbr: "IA" },
  { name: "Kansas", abbr: "KS" },
  { name: "Kentucky", abbr: "KY" },
  { name: "Louisiana", abbr: "LA" },
  { name: "Maine", abbr: "ME" },
  { name: "Maryland", abbr: "MD" },
  { name: "Massachusetts", abbr: "MA" },
  { name: "Michigan", abbr: "MI" },
  { name: "Minnesota", abbr: "MN" },
  { name: "Mississippi", abbr: "MS" },
  { name: "Missouri", abbr: "MO" },
  { name: "Montana", abbr: "MT" },
  { name: "Nebraska", abbr: "NE" },
  { name: "Nevada", abbr: "NV" },
  { name: "New Hampshire", abbr: "NH" },
  { name: "New Jersey", abbr: "NJ" },
  { name: "New Mexico", abbr: "NM" },
  { name: "New York", abbr: "NY" },
  { name: "North Carolina", abbr: "NC" },
  { name: "North Dakota", abbr: "ND" },
  { name: "Ohio", abbr: "OH" },
  { name: "Oklahoma", abbr: "OK" },
  { name: "Oregon", abbr: "OR" },
  { name: "Pennsylvania", abbr: "PA" },
  { name: "Rhode Island", abbr: "RI" },
  { name: "South Carolina", abbr: "SC" },
  { name: "South Dakota", abbr: "SD" },
  { name: "Tennessee", abbr: "TN" },
  { name: "Texas", abbr: "TX" },
  { name: "Utah", abbr: "UT" },
  { name: "Vermont", abbr: "VT" },
  { name: "Virginia", abbr: "VA" },
  { name: "Washington", abbr: "WA" },
  { name: "West Virginia", abbr: "WV" },
  { name: "Wisconsin", abbr: "WI" },
  { name: "Wyoming", abbr: "WY" },
];

// Normalize park name into a DB-safe id
function normalizeId(name) {
  return name
    .toLowerCase()
    .replace(/\s+/g, "_")
    .replace(/[^a-z0-9_]/g, "");
}

// Map states to correct Wikipedia page URLs (handles redirects/special cases)
function wikiUrlForState(name) {
  const specialCases = {
    "New Jersey": "List_of_New_Jersey_state_parks",
    "New Mexico": "List_of_New_Mexico_state_parks",
    "New York": "List_of_New_York_state_parks",
    "North Carolina": "List_of_state_parks_in_North_Carolina",
    Washington: "List_of_state_parks_in_Washington_(state)",
    "West Virginia": "List_of_state_parks_in_West_Virginia",
  };
  return `https://en.wikipedia.org/wiki/${
    specialCases[name] || "List_of_state_parks_in_" + name.replace(/\s+/g, "_")
  }`;
}

// Scrape Wikipedia tables for a given state
async function getStateParks(state) {
  const url = wikiUrlForState(state);
  try {
    const res = await fetch(url);
    if (!res.ok) {
      console.warn(`Failed to fetch ${state}: ${res.status}`);
      return [];
    }

    const html = await res.text();
    const $ = cheerio.load(html);

    const parks = [];

    // Loop over all wikitable tables on the page
    $("table.wikitable").each((_, table) => {
      $(table)
        .find("tbody tr")
        .each((_, tr) => {
          const firstTd = $(tr).find("td").first();
          const parkName = firstTd.text().trim();
          if (parkName && !parkName.includes("—")) parks.push(parkName);
        });
    });

    if (parks.length === 0)
      console.warn(`No parks found for ${state} at ${url}`);

    return parks;
  } catch (err) {
    console.warn(`Failed to fetch parks for ${state}:`, err);
    return [];
  }
}

// Main function
(async function main() {
  let sqlInserts =
    "INSERT INTO badges (id, title, description, icon_svg) VALUES\n";
  const allValues = [];

  for (const { name, abbr } of states) {
    console.log(`Fetching parks for ${name}...`);
    const parks = await getStateParks(name);
    for (const park of parks) {
      const id = `${normalizeId(park)}_state_park`;
      const title = park.replace(/'/g, "''"); // escape single quotes
      const description = `Hiked 3 trails at ${title}`;
      const icon_svg = abbr;

      allValues.push(`('${id}', '${title}', '${description}', '${icon_svg}')`);
    }
  }

  sqlInserts += allValues.join(",\n") + ";";

  fs.writeFileSync("state_parks_inserts.sql", sqlInserts, "utf-8");
  console.log(
    `Done! Generated ${allValues.length} state park badges in state_parks_inserts.sql`,
  );
})();
